### PR TITLE
Implement SDPService as a service candidate

### PR DESCRIPTION
**Pull request by lepiolet**
## Description

This PR implements a new a new CRD SDPService.
This SDPService is a ServiceCandidate that can be used to define Services in the cluster.

Right now we have Deployment and SDPService as Service Candidates but in the future it will be maybe a good idea to have only SDPService and be explicit with the service definition (or making the other entities to create SDPService automatically)

## Checklist
- [x] Bump `.version` and `.appVersion` in [k8s/crd/Chart.yaml](../k8s/crd/Chart.yaml)
- [x] Bump `.version` and `.appVersion` in [k8s/chart/Chart.yaml](../k8s/chart/Chart.yaml)

### CODE REVIEW

**Review by thomascellerier — APPROVED** [review]
We discussed this on the side.
Looks like a good approach but we will have to iterate on it.
Probably we should take in consideration the kind and not just namespace and name.
So the service id could be `namespace_kind_name` rather than the current `namespace_name`.
This will avoid conflicts with different entities of different kinds but the same name.

File: sdp-common/src/service.rs [hunk 1]
```diff
@@ -35,6 +36,51 @@ impl ToString for SDPInjectionStrategy {
     }
 }
 
+#[derive(Clone, Debug)]
+pub enum ServiceCandidate {
+    Deployment(Deployment),
+    SDPService(SDPService),
+}
```
Comment (mandopaloooza) [inline sdp-common/src/service.rs:43]: I made the [exact same wrapper in the PR to support Jobs](https://github.com/appgate/sdp-k8s-injector/pull/90/files#diff-14cae182f875cc5eb039bf8e309b6b678c7bec42d8be44f55837f77870e914ddR105-R108). I spent some time trying to find a way so we can pass the trait to the messaging queue but couldn't find one.

File: sdp-device-id-service/src/device_id_manager.rs [hunk 2]
```diff
@@ -112,48 +115,67 @@ impl DeviceIdAPI for KubeDeviceIdManager {
...
+                            rs.meta()
+                                .owner_references
+                                .as_ref()
+                                .map(|owners| owners[0].name == *owner_name)
+                                .unwrap_or(false)
+                        })
+                        .next();
+                    let num_replicas = owned_rs
+                        .and_then(|rs| rs.spec.as_ref())
+                        .and_then(|spec| spec.replicas);
```
Comment (mandopaloooza) [inline sdp-device-id-service/src/device_id_manager.rs:155]: Something to consider in the future is what to do with services that don't have replicaset as an intermediate. For example, Job doesn't own any replicasets.

File: sdp-identity-service/src/service_candidate_watcher.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+use sdp_common::annotations::SDP_INJECTOR_LABEL;
+use sdp_common::crd::SDPService;
+use sdp_common::service::ServiceCandidate;
+use sdp_common::traits::MaybeService;
+use sdp_common::watcher::SimpleWatchingProtocol;
+use sdp_common::{crd::ServiceIdentity, traits::Candidate};
+use sdp_macros::{service_candidate_protocol, when_ok};
+
+use crate::identity_manager::IdentityManagerProtocol;
+
```
Comment (mandopaloooza) [inline sdp-identity-service/src/service_candidate_watcher.rs:14]:
```suggestion

logger!("SDPServiceWatcher");

```

File: sdp-identity-service/src/service_candidate_watcher.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            fn key(&self) -> Option<String> {
+                ServiceCandidateProtocol::key(self)
+            }
+        }
+    };
+}
+
+service_candidate_protocol!(SDPService, ServiceCandidate::SDPService);
+service_candidate_protocol!(Deployment, ServiceCandidate::Deployment);
```
Comment (mandopaloooza) [inline sdp-identity-service/src/service_candidate_watcher.rs:141]: Nice! Good use of macros.

File: sdp-macros/src/lib.rs [hunk 5]
```diff
@@ -359,3 +373,48 @@ macro_rules! device_id {
...
+                ns: Option<Namespace>,
+            ) -> Option<IdentityManagerProtocol<$t, ServiceIdentity>> {
+                ServiceCandidateProtocol::deleted(self, ns)
+            }
+
+            fn key(&self) -> Option<String> {
+                ServiceCandidateProtocol::key(self)
+            }
+        }
+    };
```
Comment (mandopaloooza) [inline sdp-macros/src/lib.rs:419]: Combine this variant with the macro `service_candidate_protocol!` defined in `sdp-identity-service/src/service_candidate_watcher.rs`? Or bring that one here?

File: sdp-device-id-service/src/device_id_manager.rs (hunk 2 above)
Comment (lepiolet) [reply sdp-device-id-service/src/device_id_manager.rs:155]: I think at some point we can make this service completely dynamic

File: sdp-identity-service/src/service_candidate_watcher.rs (hunk 3 above)
Comment (lepiolet) [reply sdp-identity-service/src/service_candidate_watcher.rs:14]: ahh yes, I forgot

File: sdp-macros/src/lib.rs (hunk 5 above)
Comment (lepiolet) [reply sdp-macros/src/lib.rs:419]: mmm what do you mean exactly?

**Pull request merged**